contracts/nft_royalty/lib.rs

```diff
@@ -0,0 +1,437 @@
+            Ok(venue_id)
+        }
+
+        /// Returns the [`VenueId`] if the contract is initialized. Otherwise, returns an error.
+        #[ink(message)]
+        pub fn venue_id(&self) -> Result<VenueId> {
+            self.venue_id.ok_or(Error::MissingContractInitialization)
+        }
+
+        /// Update the `polymesh-ink` API using the tracker.
+        ///
+        /// Anyone can pay the gas fees to do the update using the tracker.
+        #[ink(message)]
+        pub fn update_polymesh_ink(&mut self) -> Result<()> {
+            self.api.check_for_upgrade()?;
+            Ok(())
+        }
+
+        /// Returns the [`IdentityId`] of the contract.
+        #[ink(message)]
+        pub fn contracts_identity(&self) -> IdentityId {
+            self.contract_identity
+        }
+
+        /// Returns the [`PortfolioId`] of the contract's caller.
+        #[ink(message)]
+        pub fn royalty_portfolio_identity(&self) -> Result<PortfolioId> {
+            let callers_identity = Self::get_callers_identity()?;
+            self.royalty_portfolio(callers_identity)
+        }
+
+        /// Returns the [`Perbill`] that corresponds to the percentage amount that the artist receives as royalty for each NFT transfer.
+        #[ink(message)]
+        pub fn royalty_percentage(&mut self, ticker: Ticker) -> Result<Perbill> {
+            let nft_artist_rules = self.decoded_asset_metadata_value(ticker)?;
+            Ok(nft_artist_rules.royalty_percentage)
+        }
+
+        /// Returns the decoded metadata value ([`NFTArtistRules`]) for the given [`Ticker`].
+        #[ink(message)]
+        pub fn decoded_asset_metadata_value(&mut self, ticker: Ticker) -> Result<NFTArtistRules> {
+            let asset_metadata_value = self.asset_metadata_value(ticker)?;
+            NFTArtistRules::decode::<&[u8]>(&mut asset_metadata_value.0.as_ref())
+                .map_err(|e| Error::FailedToDecodeMetadataValue(e.to_string()))
+        }
+
+        /// Adds a settlement instruction.
+        ///
+        /// The instruction will have three legs. One [`Leg`] where [`NFTTransferDetails::nft_owner_portfolio`] is transferring
+        /// [`NFTTransferDetails::nfts`] to [`NFTTransferDetails::nft_receiver_portfolio`], another leg where
+        /// [`NFTOffer::payer_portfolio`] sends [`NFTOffer::transfer_price`] to [`NFTOffer::receiver_portfolio`], and one leg
+        /// where the payer is transferring the royalty to the artist.
+        ///
+        /// Note: Call `royalty_percentage` to figure out the royalty percentage.
+        #[ink(message)]
+        pub fn create_transfer(
+            &mut self,
+            nft_transfer_details: NFTTransferDetails,
+            nft_offer: NFTOffer,
+        ) -> Result<Vec<Leg>> {
+            debug_println!("Creating an NFT transfer");
+            let venue_id = self.venue_id()?;
+            debug_println!("Retrieving contract's identity");
+            self.contract_identity = PolymeshInk::get_our_did()?;
```

Comment:
I  don't think we need to support the contract's identity changing after initialization.

Reply:
I have removed this, but what if the contract is initialized and then the contract's identity changes (e.g the `create_child_identity` extrisinc is called after initialization)? I also think this is an issue for the venue owner if we only allow `initialize_contract` to be called once.

Reply:
All of the portfolios would have the wrong custodian (old identity).  So this contract shouldn't be moved to a different identity.  I think in general deployed contracts shouldn't have their identity changed on them.  Supporting that would be difficult.

Reply:
To use child identities I would recommend this workflow:

1. Deploy contract (not initialized).
2. Call `create_child_identity` with the contract's address to move the contract into it's own identity as the primary key.
3. Call `initialize_contract`.